### PR TITLE
Combined merge

### PR DESCRIPTION
**Pull request by macumber**

### CODE REVIEW

File: lib/urbanopt/geojson/geo_file.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+require 'json'
+
+
+module URBANopt
+  module GeoJSON
+    class GeoFile < URBANopt::Core::FeatureFile
+      @@geojson_schema = nil
+      @@schema_file_lock = Mutex.new
+
+      def initialize(path_or_file_or_hash, runner)
```
Comment (macumber) [inline lib/urbanopt/geojson/geo_file.rb]: @karinamzalez this is too many options for the constructor.  Let's just make this take a path (as a String) to the file on disk.  Let's remove runner from all methods.  The methods can throw or return errors, we can catch these in the measure and feed to the runner there as needed.  The runner should not be used outside of the measure

File: spec/tests/urbanopt/geojson/geo_file_spec.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+#*********************************************************************************
+
+require_relative '../../../spec_helper'
+
+RSpec.describe URBANopt::GeoJSON do
+  before(:each) do
+    @spec_files_dir = File.join(File.dirname(__FILE__), '..', '..', '..', 'files')
+    @runner = OpenStudio::Measure::OSRunner.new(OpenStudio::WorkflowJSON.new)
+    @geofile = URBANopt::GeoJSON::GeoFile.new("/Users/karinamzalez/workspace/nrel/urbanopt-geojson-gem/spec/files/nrel_stm_footprints.geojson", @runner)
```
Comment (macumber) [inline spec/tests/urbanopt/geojson/geo_file_spec.rb]: @karinamzalez this path is on your computer only

File: urbanopt-geojson-gem.gemspec [hunk 3]
```diff
@@ -11,7 +11,10 @@ Gem::Specification.new do |spec|
```
Comment (macumber) [inline urbanopt-geojson-gem.gemspec]: @karinamzalez and @bgschiller you should add yourselves as authors

File: lib/urbanopt/geojson/geo_file.rb (hunk 1 above)
Comment (bgschiller) [reply lib/urbanopt/geojson/geo_file.rb]: @macumber The `path_or_file_or_hash` is a change that I made over in https://github.com/urbanopt/urbanopt-geojson-gem/pull/8#discussion_r280118835 in order to support the way urbanopt-web needs to use GeoFile. The geojson in urbanopt comes from the database, so there _isn't_ a path we could use. If we require that the GeoFile constructor accept only a `path`, we'll have to write out temporary files whenever we create one—not ideal.

I would recommend, if we want to only support a single variation, that we make it the hash option. It's the most general, meaning that wherever the data came from we can easily get it as a hash. By leaving the caller in charge of instantiating the hash, we accomplish several goals:

- `Decoupling` We don't _want_ to be the ones validating the path or deciding what encoding to use for reading the file. Those involve issues we don't know how to recover from, so they're better left to calling code.
- `Efficiency` The caller might already have the file open, or even already have the hash in memory. We don't need them to write to a file just so we can turn around and read it.
- `Flexibility` Some situations might involve reading the object from a database or from the POST params (this is the case in Urbanopt-web), when there's not a path we could offer. Some environments (like AWS lambda) have restrictions on being able to write to the file system.

(I borrowed many of these ideas from https://python-patterns.guide/gang-of-four/factory-method/#dodge-use-dependency-injection)

File: lib/urbanopt/geojson/geo_file.rb [hunk 4]
```diff
@@ -19,7 +19,7 @@ def initialize(path_or_file_or_hash)
         elsif path_or_file_or_hash.respond_to?(:read)
           @path = path_or_file_or_hash.respond_to(:path) ? path_or_file_or_hash.path : nil
```
Comment (macumber) [inline lib/urbanopt/geojson/geo_file.rb]: @karinamzalez I think this is still wrong, lets make this:

```
def initialize(path)
  raise "Incorrect path class #{path.class}" unless path.is_a?(String)
```

Comment (macumber) [reply lib/urbanopt/geojson/geo_file.rb]: anything using respond_to is too complicated :-)

**Pull request merged**